File: apps/visualizer/backend/api/providers.py
"""Providers API — list providers, models, manage fallback order."""

import sqlite3

from database import add_user_model, delete_user_model, get_user_models
from flask import Blueprint, current_app, jsonify, request
from spectree import Response

from api.openapi import spec
from api.schemas.jobs import ErrorBody
from api.schemas.providers import (
    DescriptionModelsResponse,
    FallbackOrderResponse,
    Provider,
    ProviderDefaults,
    ProviderDeletedResponse,
    ProviderHealthResponse,
    ProviderModel,
    ProviderListResponse,
    ProviderModelsListResponse,
    ProviderReorderSuccessResponse,
)
from lightroom_tagger.core.provider_registry import ProviderRegistry

bp = Blueprint("providers", __name__)


def _get_registry() -> ProviderRegistry:
    return ProviderRegistry()


@bp.route("/", methods=["GET"])
@spec.validate(
    resp=Response(HTTP_200=ProviderListResponse),
    tags=['providers'],
)
def list_providers():
    registry = _get_registry()
    return jsonify(registry.list_providers())


@bp.route("/fallback-order", methods=["GET", "PUT"])
@spec.validate(
    resp=Response(HTTP_200=FallbackOrderResponse, HTTP_400=ErrorBody),
    tags=['providers'],
)
def fallback_order():
    registry = _get_registry()
    if request.method == "GET":
        return jsonify({"order": registry.fallback_order})
    data = request.json
    if not data or "order" not in data:
        return jsonify({"error": "order is required"}), 400
    try:
        registry.update_fallback_order(data["order"])
    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    return jsonify({"order": registry.fallback_order})


@bp.route("/defaults", methods=["GET", "PUT"])
@spec.validate(
    resp=Response(HTTP_200=ProviderDefaults, HTTP_400=ErrorBody),
    tags=['providers'],
)
def defaults():
    registry = _get_registry()
    if request.method == "GET":
        return jsonify(registry.defaults)
    data = request.json
    if not data:
        return jsonify({"error": "body is required"}), 400
    try:
        registry.update_defaults(data)
    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    return jsonify(registry.defaults)


@bp.route("/models/description", methods=["GET"])
@spec.validate(
    resp=Response(HTTP_200=DescriptionModelsResponse),
    tags=['providers'],
)
def all_description_models():
    """Flat list of all models across all providers, for the NL/description task selector.

    Each model entry includes ``tool_calling`` from the provider's ``providers.json``
    config (function-calling / tools support).

    For providers with no statically configured models (e.g. oMLX), attempts a live
    /v1/models discovery with a short timeout so the selector still includes them when
    they are running.
    """
    registry = _get_registry()
    result = []
    for provider in registry.list_providers():
        pid = provider["id"]
        models_list = registry.list_models(pid)

        if not models_list:
            try:
                client = registry.get_client(pid)
                discovered = list(client.models.list(timeout=2.0))
                models_list = [
                    {"id": m.id, "name": m.id, "source": "discovered"}
                    for m in discovered
                ]
            except Exception:
                pass

        seen = set()
        for m in models_list:
            key = (pid, m["id"])
            if key in seen:
                continue
            seen.add(key)
            result.append({
                "provider_id": pid,
                "provider_name": provider["name"],
                "model_id": m["id"],
                "model_name": m.get("name", m["id"]),
                "tool_calling": bool(provider.get("tool_calling", False)),
            })

    defaults = registry.defaults.get("description", {}) or {}
    return jsonify({
        "models": result,
        "default_provider": defaults.get("provider"),
        "default_model": defaults.get("model"),
    })


@bp.route("/<provider_id>/health", methods=["GET"])
@spec.validate(
    resp=Response(HTTP_200=ProviderHealthResponse, HTTP_404=ErrorBody),
    tags=['providers'],
)
def provider_health(provider_id: str):
    registry = _get_registry()
    try:
        ok, detail = registry.probe_connection(provider_id)
    except KeyError:
        return jsonify({"error": "Unknown provider"}), 404
    if ok:
        return jsonify({"reachable": True})
    return jsonify({"reachable": False, "error": detail})


@bp.route("/<provider_id>/models/<path:model_id>", methods=["DELETE"])
@spec.validate(
    resp=Response(HTTP_200=ProviderDeletedResponse, HTTP_404=ErrorBody),
    tags=['providers'],
)
def delete_model(provider_id: str, model_id: str):
    deleted = delete_user_model(current_app.db, provider_id, model_id)
    if deleted:
        return jsonify({"deleted": True})
    registry = _get_registry()
    try:
        removed = registry.remove_model(provider_id, model_id)
    except KeyError:
        return jsonify({"error": f"Unknown provider: {provider_id}"}), 404
    if removed:
        return jsonify({"deleted": True})
    return jsonify({"error": "Model not found"}), 404


@bp.route("/<provider_id>/models/order", methods=["PUT"])
@spec.validate(
    resp=Response(HTTP_200=ProviderReorderSuccessResponse, HTTP_400=ErrorBody, HTTP_404=ErrorBody),
    tags=['providers'],
)
def reorder_models_endpoint(provider_id: str):
    """Reorder models for a provider."""
    registry = _get_registry()
    data = request.json
    if not data or "order" not in data:
        return jsonify({"error": "order is required"}), 400
    try:
        registry.reorder_models(provider_id, data["order"])
        return jsonify({"success": True})
    except KeyError:
        return jsonify({"error": f"Unknown provider: {provider_id}"}), 404
    except Exception as e:
        return jsonify({"error": str(e)}), 400


@bp.route("/<provider_id>/models", methods=["GET", "POST"])
@spec.validate(
    resp=Response(
        HTTP_200=ProviderModelsListResponse,
        HTTP_201=ProviderModel,
        HTTP_400=ErrorBody,
        HTTP_404=ErrorBody,
        HTTP_409=ErrorBody,
    ),
    tags=['providers'],
)
def models(provider_id: str):
    registry = _get_registry()
    if provider_id not in [provider["id"] for provider in registry.list_providers()]:
        return jsonify({"error": f"Unknown provider: {provider_id}"}), 404

    if request.method == "GET":
        models_list = registry.list_models(provider_id)
        existing_model_ids = {model_entry["id"] for model_entry in models_list}
        user_models = get_user_models(current_app.db, provider_id)
        for user_model in user_models:
            user_model_id = user_model["model_id"]
            if user_model_id in existing_model_ids:
                continue
            existing_model_ids.add(user_model_id)
            models_list.append({
                "id": user_model_id,
                "name": user_model["model_name"],
                "vision": bool(user_model["vision"]),
                "source": "user",
            })
        
        # Apply custom order if it exists
        provider_config = registry._providers.get(provider_id, {})
        model_order = provider_config.get("model_order", [])
        if model_order:
            models_by_id = {m["id"]: m for m in models_list}
            ordered = []
            for model_id in model_order:
                if model_id in models_by_id:
                    ordered.append(models_by_id[model_id])
            # Add any models not in the order (new discoveries)
            for model in models_list:
                if model["id"] not in model_order:
                    ordered.append(model)
            models_list = ordered
        
        return jsonify(models_list)

    data = request.json
    if not data:
        return jsonify({"error": "id and name are required"}), 400

    model_id = data.get("id")
    model_name = data.get("name")
    if not isinstance(model_id, str) or not model_id.strip():
        return jsonify({"error": "id must be a non-empty string"}), 400
    if not isinstance(model_name, str) or not model_name.strip():
        return jsonify({"error": "name must be a non-empty string"}), 400

    vision = data.get("vision", True)
    if "vision" in data and not isinstance(vision, bool):
        return jsonify({"error": "vision must be a boolean"}), 400

    try:
        add_user_model(
            current_app.db,
            provider_id,
            model_id,
            model_name,
            vision,
        )
    except sqlite3.IntegrityError:
        return jsonify(
            {"error": f"Model {model_id} already exists for {provider_id}"}
        ), 409
    return (
        jsonify({
            "id": model_id,
            "name": model_name,
            "vision": vision,
            "source": "user",
        }),
        201,
    )

File: apps/visualizer/backend-ts/src/api/providers.ts
/**
 * Providers API — list providers, manage models, fallback order and defaults.
 *
 * A fresh `ProviderRegistry` per request, matching `_get_registry()`. That is not
 * an oversight: `providers.json` is user-editable on disk, and a long-lived
 * registry would keep serving a stale copy after the user edited it.
 *
 * These routes read `providers.json` and `visualizer.db`, never `library.db`, so
 * they do not use the library middleware.
 */
import { createRoute, z } from '@hono/zod-openapi';
import { existsSync } from 'node:fs';
import { config } from '../config.js';
import { openDb, type Db } from '../db/connection.js';
import {
  addUserModel,
  deleteUserModel,
  DuplicateUserModelError,
  getUserModels,
} from '../db/jobs/user-models.js';
import {
  applyModelOrder,
  ProviderRegistry,
  UnknownProviderError,
  type ProviderModelEntry,
} from '../providers/registry.js';
import { createOpenApiApp } from './openapi.js';
import { jsonBody, withValidationError } from './route-helpers.js';
import { ErrorBody } from './schemas/errors.js';
import {
  DescriptionModelsResponse,
  FallbackOrderResponse,
  ProviderDefaults,
  ProviderDeletedResponse,
  ProviderHealthResponse,
  ProviderListResponse,
  ProviderModel,
  ProviderModelsListResponse,
  ProviderReorderSuccessResponse,
} from './schemas/providers.js';

export const providersRoutes = createOpenApiApp();

/**
 * Open `visualizer.db` for a single operation.
 *
 * Not middleware: only three of the eight provider routes touch it, and the
 * middleware would have to be mounted on a prefix shared with the ones that do not.
 */
function withJobsDb<T>(write: boolean, fn: (db: Db) => T): T {
  const db = openDb(config.VISUALIZER_DB, { readonly: !write, vec: false });
  try {
    return fn(db);
  } finally {
    db.close();
  }
}

/** `request.json` is `None` for an absent or unparseable body. */
async function readJson(c: { req: { json: () => Promise<unknown> } }): Promise<unknown> {
  try {
    return await c.req.json();
  } catch {
    return null;
  }
}

// --- providers --------------------------------------------------------------

const listRoute = createRoute({
  method: 'get',
  path: '/providers/',
  tags: ['providers'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(ProviderListResponse) },
  }),
});

providersRoutes.openapi(listRoute, async (c) =>
  c.json(await new ProviderRegistry().listProviders(), 200),
);

const fallbackOrderGetRoute = createRoute({
  method: 'get',
  path: '/providers/fallback-order',
  tags: ['providers'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(FallbackOrderResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
  }),
});

providersRoutes.openapi(fallbackOrderGetRoute, (c) =>
  c.json({ order: new ProviderRegistry().fallbackOrder }, 200),
);

const fallbackOrderPutRoute = createRoute({
  method: 'put',
  path: '/providers/fallback-order',
  tags: ['providers'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(FallbackOrderResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
  }),
});

providersRoutes.openapi(fallbackOrderPutRoute, async (c) => {
  const registry = new ProviderRegistry();
  const data = await readJson(c);
  if (!data || typeof data !== 'object' || !('order' in data)) {
    return c.json({ error: 'order is required' }, 400);
  }
  try {
    registry.updateFallbackOrder((data as { order: unknown }).order);
  } catch (e) {
    if (e instanceof RangeError) return c.json({ error: e.message }, 400);
    throw e;
  }
  return c.json({ order: registry.fallbackOrder }, 200);
});

const defaultsGetRoute = createRoute({
  method: 'get',
  path: '/providers/defaults',
  tags: ['providers'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(ProviderDefaults) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
  }),
});

providersRoutes.openapi(defaultsGetRoute, (c) =>
  c.json(new ProviderRegistry().defaults as z.infer<typeof ProviderDefaults>, 200),
);

const defaultsPutRoute = createRoute({
  method: 'put',
  path: '/providers/defaults',
  tags: ['providers'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(ProviderDefaults) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
  }),
});

providersRoutes.openapi(defaultsPutRoute, async (c) => {
  const registry = new ProviderRegistry();
  const data = await readJson(c);
  if (!data) return c.json({ error: 'body is required' }, 400);
  try {
    registry.updateDefaults(data);
  } catch (e) {
    if (e instanceof RangeError) return c.json({ error: e.message }, 400);
    throw e;
  }
  return c.json(registry.defaults as z.infer<typeof ProviderDefaults>, 200);
});

// --- description model selector ---------------------------------------------

const descriptionModelsRoute = createRoute({
  method: 'get',
  path: '/providers/models/description',
  tags: ['providers'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(DescriptionModelsResponse) },
  }),
});

/**
 * Every model across every provider, flattened for the description task selector.
 *
 * Providers with no statically configured models (oMLX, for one) get a live
 * `/v1/models` discovery with a short timeout, so the selector still lists them
 * while they are running and simply omits them when they are not.
 */
providersRoutes.openapi(descriptionModelsRoute, async (c) => {
  const registry = new ProviderRegistry();
  const result: z.infer<typeof DescriptionModelsResponse>['models'] = [];

  for (const provider of await registry.listProviders()) {
    const pid = provider.id;
    let modelsList = await registry.listModels(pid);

    if (modelsList.length === 0) {
      try {
        modelsList = (await registry.discoverOpenAiModels(pid, 2000)).map((id) => ({
          id,
          name: id,
          source: 'discovered' as const,
        }));
      } catch {
        // Not running, or no such endpoint. Leave it out of the selector.
      }
    }

    const seen = new Set<string>();
    for (const m of modelsList) {
      if (seen.has(m.id)) continue;
      seen.add(m.id);
      result.push({
        provider_id: pid,
        provider_name: provider.name,
        model_id: m.id,
        model_name: m.name ?? m.id,
        tool_calling: Boolean(provider.tool_calling),
      });
    }
  }

  const defaults = registry.defaults.description ?? { provider: undefined, model: undefined };
  return c.json(
    {
      models: result,
      default_provider: defaults.provider ?? null,
      default_model: defaults.model ?? null,
    },
    200,
  );
});

// --- per-provider -----------------------------------------------------------

const healthRoute = createRoute({
  method: 'get',
  path: '/providers/{provider_id}/health',
  tags: ['providers'],
  request: { params: z.object({ provider_id: z.string() }) },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(ProviderHealthResponse) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

providersRoutes.openapi(healthRoute, async (c) => {
  const registry = new ProviderRegistry();
  let probe;
  try {
    probe = await registry.probeConnection(c.req.param('provider_id'));
  } catch (e) {
    if (e instanceof UnknownProviderError) return c.json({ error: 'Unknown provider' }, 404);
    throw e;
  }
  // Unreachable is a 200 with `reachable: false` — the provider being down is a
  // state the UI renders, not a request failure.
  return probe.ok
    ? c.json({ reachable: true }, 200)
    : c.json({ reachable: false, error: probe.detail }, 200);
});

const modelsOrderRoute = createRoute({
  method: 'put',
  path: '/providers/{provider_id}/models/order',
  tags: ['providers'],
  request: { params: z.object({ provider_id: z.string() }) },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(ProviderReorderSuccessResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

providersRoutes.openapi(modelsOrderRoute, async (c) => {
  const registry = new ProviderRegistry();
  const data = await readJson(c);
  if (!data || typeof data !== 'object' || !('order' in data)) {
    return c.json({ error: 'order is required' }, 400);
  }
  try {
    registry.reorderModels(c.req.param('provider_id'), (data as { order: unknown }).order);
    return c.json({ success: true }, 200);
  } catch (e) {
    if (e instanceof UnknownProviderError) return c.json({ error: e.message }, 404);
    return c.json({ error: e instanceof Error ? e.message : String(e) }, 400);
  }
});

// --- model list and creation ------------------------------------------------

const modelsListRoute = createRoute({
  method: 'get',
  path: '/providers/{provider_id}/models',
  tags: ['providers'],
  request: { params: z.object({ provider_id: z.string() }) },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(ProviderModelsListResponse) },
    201: { description: 'Created', content: jsonBody(ProviderModel) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
    409: { description: 'Conflict', content: jsonBody(ErrorBody) },
  }),
});

providersRoutes.openapi(modelsListRoute, async (c) => {
  const registry = new ProviderRegistry();
  const providerId = c.req.param('provider_id');
  if (!registry.hasProvider(providerId)) {
    return c.json({ error: `Unknown provider: ${providerId}` }, 404);
  }

  const modelsList: ProviderModelEntry[] = await registry.listModels(providerId);
  const existing = new Set(modelsList.map((m) => m.id));
  const userModels = existsSync(config.VISUALIZER_DB)
    ? withJobsDb(false, (db) => getUserModels(db, providerId))
    : [];
  for (const userModel of userModels) {
    if (existing.has(userModel.model_id)) continue;
    existing.add(userModel.model_id);
    modelsList.push({
      id: userModel.model_id,
      name: userModel.model_name,
      vision: Boolean(userModel.vision),
      source: 'user',
    });
  }

  // The custom order is applied a second time here, after the user models were
  // appended — `listModels` already ordered the config and discovered ones, and
  // this pass folds the user additions into the same sequence.
  return c.json(
    applyModelOrder(modelsList, registry.modelOrderFor(providerId)) as z.infer<
      typeof ProviderModelsListResponse
    >,
    200,
  );
});

const modelsCreateRoute = createRoute({
  method: 'post',
  path: '/providers/{provider_id}/models',
  tags: ['providers'],
  request: { params: z.object({ provider_id: z.string() }) },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(ProviderModelsListResponse) },
    201: { description: 'Created', content: jsonBody(ProviderModel) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
    409: { description: 'Conflict', content: jsonBody(ErrorBody) },
  }),
});

providersRoutes.openapi(modelsCreateRoute, async (c) => {
  const registry = new ProviderRegistry();
  const providerId = c.req.param('provider_id');
  if (!registry.hasProvider(providerId)) {
    return c.json({ error: `Unknown provider: ${providerId}` }, 404);
  }

  const data = await readJson(c);
  if (!data || typeof data !== 'object') {
    return c.json({ error: 'id and name are required' }, 400);
  }
  const body = data as Record<string, unknown>;
  const modelId = body.id;
  const modelName = body.name;
  if (typeof modelId !== 'string' || !modelId.trim()) {
    return c.json({ error: 'id must be a non-empty string' }, 400);
  }
  if (typeof modelName !== 'string' || !modelName.trim()) {
    return c.json({ error: 'name must be a non-empty string' }, 400);
  }
  // Defaults to true, but an explicit non-boolean is rejected rather than coerced.
  if ('vision' in body && typeof body.vision !== 'boolean') {
    return c.json({ error: 'vision must be a boolean' }, 400);
  }
  const vision = 'vision' in body ? (body.vision as boolean) : true;

  try {
    withJobsDb(true, (db) => addUserModel(db, providerId, modelId, modelName, vision));
  } catch (e) {
    if (e instanceof DuplicateUserModelError) return c.json({ error: e.message }, 409);
    throw e;
  }
  return c.json({ id: modelId, name: modelName, vision, source: 'user' as const }, 201);
});

// --- model deletion ---------------------------------------------------------

/**
 * Delete a model — the user-added row first, then the `providers.json` entry.
 *
 * The order matters: a config model and a user model can share an id, and the user
 * row is the one the user just added, so it is the one they mean.
 */
async function deleteModel(
  providerId: string,
  modelId: string,
): Promise<{ status: 200 | 404; body: { deleted: true } | { error: string } }> {
  const deleted = existsSync(config.VISUALIZER_DB)
    ? withJobsDb(true, (db) => deleteUserModel(db, providerId, modelId))
    : false;
  if (deleted) return { status: 200, body: { deleted: true } };

  const registry = new ProviderRegistry();
  try {
    if (registry.removeModel(providerId, modelId)) {
      return { status: 200, body: { deleted: true } };
    }
  } catch (e) {
    if (e instanceof UnknownProviderError) {
      return { status: 404, body: { error: `Unknown provider: ${providerId}` } };
    }
    throw e;
  }
  return { status: 404, body: { error: 'Model not found' } };
}

const modelDeleteRoute = createRoute({
  method: 'delete',
  path: '/providers/{provider_id}/models/{model_id}',
  tags: ['providers'],
  request: { params: z.object({ provider_id: z.string(), model_id: z.string() }) },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(ProviderDeletedResponse) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

/**
 * Registered twice, and this one is load-bearing rather than defensive.
 *
 * Model ids contain slashes — `meta/llama-4-maverick-17b-128e-instruct`,
 * `google/gemini-2.5-flash-lite` — which is why Flask declared the segment with the
 * `path:` converter. `@hono/zod-openapi` derives the Hono pattern from the OpenAPI
 * path template, and `{model_id}` becomes a single-segment `:model_id`, so a slashed
 * id would 404.
 *
 * The greedy plain route below handles every request; the `openapi()` registration
 * is what puts the path in the document. Both call the same function, so there is
 * one implementation and no way for them to disagree.
 */
providersRoutes.openapi(modelDeleteRoute, async (c) => {
  const result = await deleteModel(c.req.param('provider_id'), c.req.param('model_id'));
  return result.status === 200
    ? c.json(result.body as { deleted: true }, 200)
    : c.json(result.body as { error: string }, 404);
});

providersRoutes.delete('/providers/:provider_id/models/:model_id{.+}', async (c) => {
  const result = await deleteModel(c.req.param('provider_id'), c.req.param('model_id'));
  return c.json(result.body, result.status);
});
